### PR TITLE
builder-context: Write a gitignore file into the state-dir

### PR DESCRIPTION
**Pull request by alatiera**
Unless specified, state-dir is placed on the current path, which is usually the source directory of the project.

Write a .gitignore file to make sure that both the state-dir never gets checked out into git and that all the tooling will ignore it and avoid indexing it.

Inpired by meson's gitignore file in the build directories.

Retro-actively discovered the same change in flatpak-build init [1]

[1]: https://github.com/flatpak/flatpak/commit/a62f8cfb62c4232ef2674cfa4431169a292120d6

### CODE REVIEW

File: src/builder-context.c [hunk 1]
```diff
@@ -1193,6 +1193,58 @@ builder_context_get_sdk_config (BuilderContext *self)
...
+{
+  g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+
+  gboolean exists = g_file_query_exists (self->state_dir, NULL);
+
+  if (!exists && !g_file_make_directory_with_parents (self->state_dir, NULL, error))
+  {
+    flatpak_fail (error, "Could not create state dir");
+    return FALSE;
+  }
```
Comment (TingPing) [inline src/builder-context.c:1208]:
```suggestion
  if (!g_file_make_directory_with_parents (self->state_dir, NULL, error))
  {
    if (g_error_matches(error, G_IO_ERROR, G_IO_ERROR_EXISTS))
      return TRUE;

    flatpak_fail (error, "Could not create state dir");
    return FALSE;
  }
```

Race condition between check and creation.

File: src/builder-context.c [hunk 2]
```diff
@@ -1193,6 +1193,58 @@ builder_context_get_sdk_config (BuilderContext *self)
...
+  // created before or it was deliberately modified or deleted
+  if (!exists) {
+    g_autoptr (GFile) gitignore_file = g_file_get_child (self->state_dir, ".gitignore");
+
+    if (!g_file_query_exists (gitignore_file, NULL)) {
+        const char *git_ignore_file = "# This file is autogenerated by flatpak-builder.\n"
+          "# If you change or delete it, it won't be recreated. \n"
+          "*\n";
+
+        g_autoptr(GFileOutputStream) out = g_file_create (gitignore_file, G_FILE_CREATE_NONE, NULL, error);
```
Comment (TingPing) [inline src/builder-context.c:1221]: Again don't check if it exists. Just write it. Then handle `G_IO_ERROR_EXISTS`.

**Comment by alatiera** [comment]
Indeed, hit the footguns. I guess that's why everyone is using `g_file_replace_contents` even for supposedly simple things? Should I just make it a one liner `g_file_replace_contents` and give up on the `If you change or delete it, it won't be recreated.` part, it would match the code in flatpak itself anyway.

**Comment by TingPing** [comment]
Yes that sounds fine to me.

**Comment by alatiera** [comment]
(pushed but want to test it again)

**Comment by alatiera** [comment]
Yea the current version did break something.

```
(/home/alatiera/Projects/flatpak-builder/_builddir_vscode/src/flatpak-builder:34220): GLib-CRITICAL **: 23:46:01.803: g_file_set_contents_full: assertion 'error == NULL || *error == NULL' failed

Thread 1 "flatpak-builder" received signal SIGTRAP, Trace/breakpoint trap.
_g_log_abort (breakpoint=<optimized out>) at ../glib/gmessages.c:426
426         G_BREAKPOINT ();
(gdb) bt
#0  _g_log_abort (breakpoint=<optimized out>) at ../glib/gmessages.c:426
#1  g_logv (log_domain=0x7ffff7f0d00e "GLib", log_level=G_LOG_LEVEL_CRITICAL, format=<optimized out>, args=args@entry=0x7fffffffdac0)
    at ../glib/gmessages.c:1273
#2  0x00007ffff7eb1b3b in g_log
    (log_domain=log_domain@entry=0x7ffff7f0d00e "GLib", log_level=log_level@entry=G_LOG_LEVEL_CRITICAL, format=format@entry=0x7ffff7f0f2ec "%s: assertion '%s' failed") at ../glib/gmessages.c:1315
#3  0x00007ffff7eb2c4d in g_return_if_fail_warning
    (log_domain=log_domain@entry=0x7ffff7f0d00e "GLib", pretty_function=pretty_function@entry=0x7ffff7f1eb70 <__func__.17> "g_file_set_contents_full", expression=expression@entry=0x7ffff7f127f0 "error == NULL || *error == NULL") at ../glib/gmessages.c:3055
#4  0x00007ffff7e910f2 in g_file_set_contents_full
    (filename=0x5555555e7250 "/home/alatiera/Projects/flatpak-builder/.flatpak-builder/checksums/x86_64-test-runtime.json", contents=contents@entry=0x5555555dd0f0 "7109820d75e46b6c85799365dbc67797583c15d259666e0b889e258185e67e3f", length=length@entry=-1, flags=flags@entry=(G_FILE_SET_CONTENTS_CONSISTENT | G_FILE_SET_CONTENTS_ONLY_EXISTING), mode=mode@entry=438, error=error@entry=0x7fffffffdd50)
    at ../glib/gfileutils.c:1337
#5  0x00007ffff7e91447 in g_file_set_contents
    (filename=<optimized out>, contents=contents@entry=0x5555555dd0f0 "7109820d75e46b6c85799365dbc67797583c15d259666e0b889e258185e67e3f", length=length@entry=-1, error=error@entry=0x7fffffffdd50) at ../glib/gfileutils.c:1253
#6  0x000055555556ac6f in builder_context_set_checksum_for
    (self=self@entry=0x5555555dcb90 [BuilderContext], name=name@entry=0x5555555da580 "test-runtime.json", checksum=checksum@entry=0x5555555dd0f0 "7109820d75e46b6c85799365dbc67797583c15d259666e0b889e258185e67e3f", error=error@entry=0x7fffffffdd50)
    at ../src/builder-context.c:553
#7  0x000055555556569c in main (argc=<optimized out>, argv=<optimized out>) at ../src/builder-main.c:824
```

**Comment by alatiera** [comment]
Nvm, fixed, was missing the `clear_error`

**Review by TingPing — APPROVED** [review]
> (pushed but want to test it again)

LGTM, let me know if its ready to merge.

**Comment by alatiera** [comment]
Good to go, just fixed a typo yday

**Pull request merged**